File: python/mirage/core/disk/rename.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import aiofiles.os

from mirage.accessor.disk import DiskAccessor
from mirage.cache.context import invalidate_subtree
from mirage.core.disk.utils import resolve_inside
from mirage.types import PathSpec


async def rename(accessor: DiskAccessor, src_spec: PathSpec,
                 dst_spec: PathSpec) -> None:
    src = src_spec.mount_path
    dst = dst_spec.mount_path
    root = accessor.root
    await invalidate_subtree(src_spec)
    # Both sides are subtree evictions: a rename destroys the destination's
    # previous identity and relocates everything under the source, so a
    # listing or body cached below either name is now stale.
    await invalidate_subtree(dst_spec)
    await aiofiles.os.rename(resolve_inside(root, src, src_spec),
                             resolve_inside(root, dst, dst_spec))

File: typescript/packages/node/src/core/disk/rename.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { DiskAccessor } from '../../accessor/disk.ts'
import { rename as fsRename } from 'node:fs/promises'
import { invalidateSubtree } from '@struktoai/mirage-core/cache/context'
import type { PathSpec } from '@struktoai/mirage-core/types'
import { enoent } from '@struktoai/mirage-core/utils/errors'
import { resolveInside } from './utils.ts'

export async function rename(accessor: DiskAccessor, src: PathSpec, dst: PathSpec): Promise<void> {
  const s = await resolveInside(accessor.root, src)
  const d = await resolveInside(accessor.root, dst)
  try {
    await fsRename(s, d)
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
      throw enoent(src)
    }
    throw err
  }
  await invalidateSubtree(src)
  // Both sides are subtree evictions: a rename destroys the destination's
  // previous identity and relocates everything under the source, so a
  // listing or body cached below either name is now stale.
  await invalidateSubtree(dst)
}
